pkg/storegateway/bucket_chunk_reader.go

```diff
@@ -166,7 +166,7 @@ func (r *bucketChunkReader) loadChunks(ctx context.Context, res []seriesChunks,
 			return errors.Wrap(err, "read chunk")
 		}
 
-		err = populateChunk(&(res[pIdx.seriesEntry].chks[pIdx.chunkEntry]), rawChunk(cb))
+		err = populateChunk(&(res[pIdx.seriesEntry].chks[pIdx.chunkEntry]), rawChunk{b: &cb})
```

Comment:
But using a slice ref `rawChunk` fits in the interface, as in essence is just a pointer wrapper. 

Plus, the escape analyzer tells a different story. 

before: 
```
pkg/storegateway/bucket_chunk_reader.go:163:86: pIdx.offset escapes to heap
pkg/storegateway/bucket_chunk_reader.go:163:99: pIdx.length escapes to heap
pkg/storegateway/bucket_chunk_reader.go:169:79: rawChunk(cb) escapes to heap
pkg/storegateway/bucket_chunk_reader.go:177:56: inlining call to varint.UvarintSize
pkg/storegateway/bucket_chunk_reader.go:182:101: leaking param: dest
```

after: 
```
pkg/storegateway/bucket_chunk_reader.go:163:86: pIdx.offset escapes to heap
pkg/storegateway/bucket_chunk_reader.go:163:99: pIdx.length escapes to heap
pkg/storegateway/bucket_chunk_reader.go:177:56: inlining call to varint.UvarintSize
pkg/storegateway/bucket_chunk_reader.go:182:101: leaking param: dest
```

Reply:
I am surprised by this :thinking: 

Still, the fact that you pass in a pointer to `cb` now makes `cb` escape above:
```
pkg/storegateway/bucket_chunk_reader.go:152:3: moved to heap: cb
```

Reply:
In order to prevent that from escaping, I'd suggest to have a place for that in `*bucketChunkReader`:

```diff
diff --git pkg/storegateway/bucket_chunk_reader.go pkg/storegateway/bucket_chunk_reader.go
index 57efdd32a..ee250ee33 100644
--- pkg/storegateway/bucket_chunk_reader.go
+++ pkg/storegateway/bucket_chunk_reader.go
@@ -34,6 +34,12 @@ type bucketChunkReader struct {
        block *bucketBlock
 
        toLoad [][]loadIdx
+
+       buf struct {
+               // c can be used to put a chunk and pass a reference to other functions.
+               // Since bucketChunkReader is going to escape anyway, this avoids each chunk escaping separately.
+               c []byte
+       }
 }
 
 func newBucketChunkReader(ctx context.Context, block *bucketBlock) *bucketChunkReader {
@@ -149,9 +155,9 @@ func (r *bucketChunkReader) loadChunks(ctx context.Context, res []seriesChunks,
                }
                // We ignore the crc32 after the chunk data.
                chunkEncDataLen := chunks.ChunkEncodingSize + int(chunkDataLen)
-               cb := chunksPool.Get(chunkEncDataLen)
+               r.buf.c = chunksPool.Get(chunkEncDataLen)
 
-               fullyRead, err := io.ReadFull(reader, cb)
+               fullyRead, err := io.ReadFull(reader, r.buf.c)
                // We get io.EOF when there are 0 bytes read and io.UnexpectedEOF when there are some but not all read.
                if errors.Is(err, io.EOF) || errors.Is(err, io.ErrUnexpectedEOF) {
                        if chunksLeft := len(pIdxs) - 1 - i; chunksLeft != 0 {
@@ -159,14 +165,14 @@ func (r *bucketChunkReader) loadChunks(ctx context.Context, res []seriesChunks,
                                // Any other errors are definitely unexpected.
                                return fmt.Errorf("underread with %d more remaining chunks in seq %d start %d end %d", chunksLeft, seq, part.Start, part.End)
                        }
-                       if err = r.fetchChunkRemainder(ctx, seq, int64(reader.offset), int64(chunkEncDataLen-fullyRead), cb[fullyRead:], &localStats); err != nil {
+                       if err = r.fetchChunkRemainder(ctx, seq, int64(reader.offset), int64(chunkEncDataLen-fullyRead), r.buf.c[fullyRead:], &localStats); err != nil {
                                return errors.Wrapf(err, "refetching chunk seq %d offset %x length %d", seq, pIdx.offset, pIdx.length)
                        }
                } else if err != nil {
                        return errors.Wrap(err, "read chunk")
                }
 
-               err = populateChunk(&(res[pIdx.seriesEntry].chks[pIdx.chunkEntry]), rawChunk{b: &cb})
+               err = populateChunk(&(res[pIdx.seriesEntry].chks[pIdx.chunkEntry]), rawChunk{b: &r.buf.c})
                if err != nil {
                        return errors.Wrap(err, "populate chunk")
                }

```

Reply:
can we just pass in a byte slice and convert it to `rawChunk`? This is a diff from `main` and the allocation is gone too. I think it's because there's no more interface conversion


```diff
diff --git a/pkg/storegateway/bucket_chunk_reader.go b/pkg/storegateway/bucket_chunk_reader.go
index 3e29d14e7..e46acc95e 100644
--- a/pkg/storegateway/bucket_chunk_reader.go
+++ b/pkg/storegateway/bucket_chunk_reader.go
@@ -166,7 +166,7 @@ func (r *bucketChunkReader) loadChunks(ctx context.Context, res []seriesChunks,
 			return errors.Wrap(err, "read chunk")
 		}
 
-		err = populateChunk(&(res[pIdx.seriesEntry].chks[pIdx.chunkEntry]), rawChunk(cb))
+		err = populateChunk(&(res[pIdx.seriesEntry].chks[pIdx.chunkEntry]), cb)
 		if err != nil {
 			return errors.Wrap(err, "populate chunk")
 		}
@@ -198,7 +198,8 @@ func (r *bucketChunkReader) fetchChunkRemainder(ctx context.Context, seq int, of
 }
 
 // populateChunk retains in.Bytes() in out.Raw.
-func populateChunk(out *storepb.AggrChunk, in chunkenc.Chunk) error {
+func populateChunk(out *storepb.AggrChunk, inBytes []byte) error {
+	in := rawChunk(inBytes)
 	var enc storepb.Chunk_Encoding
 	switch in.Encoding() {
 	case chunkenc.EncXOR:
```

Reply:
or perhaps an even simpler change has the same effect AFAICT

```diff
diff --git a/pkg/storegateway/bucket_chunk_reader.go b/pkg/storegateway/bucket_chunk_reader.go
index 3e29d14e7..07fb95bdb 100644
--- a/pkg/storegateway/bucket_chunk_reader.go
+++ b/pkg/storegateway/bucket_chunk_reader.go
@@ -198,7 +198,7 @@ func (r *bucketChunkReader) fetchChunkRemainder(ctx context.Context, seq int, of
 }
 
 // populateChunk retains in.Bytes() in out.Raw.
-func populateChunk(out *storepb.AggrChunk, in chunkenc.Chunk) error {
+func populateChunk(out *storepb.AggrChunk, in rawChunk) error {
 	var enc storepb.Chunk_Encoding
 	switch in.Encoding() {
 	case chunkenc.EncXOR:
```

Reply:
I think this last change is what we have to do, instead of having to solve our own problems.

Reply:
I also opt for that change. It's the simplest way to get rid of the problem.